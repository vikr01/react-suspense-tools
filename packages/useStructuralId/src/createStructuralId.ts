import { nanoid } from 'nanoid/non-secure';

type ElementType = string | object | Function;
type Key = string | number;

declare const __DEV_STRUCTURAL_ID_DEBUG__: boolean;

const objectHashes = new WeakMap<object, string>();

function getElementTypeId(elementType: ElementType): string {
  if (elementType == null) { return 'Unknown'; }
  if (typeof elementType !== 'object' && typeof elementType !== 'function') return elementType.toString();

  if (!objectHashes.has(elementType)) {
    let id: string;

    // This will get stripped out in prod as dead code
    if (process.env.NODE_ENV === 'development' && __DEV_STRUCTURAL_ID_DEBUG__) {
      id = (elementType as any)?.displayName ?? (elementType as any).name ?? elementType.toString();
    } else {
      id = nanoid(6);
    }

    objectHashes.set(elementType, id);
  }

  return objectHashes.get(elementType)!;
}

export function createArrayId(
  arr: ReadonlyArray<[ElementType, Key]>
): string {
  const parts: string[] = [];

  for (let i = 0; i < arr.length; i++) {
    const [elementType, key] = arr[i];
    const elementTypeId = getElementTypeId(elementType);
    const encodedKey =
      typeof key === 'string' ? `s:${key}` : `n:${key}`;
    parts.push(`${elementTypeId}:${encodedKey}`);
  }

  return parts.join(',');
}

export function createArrayIdWithNumber(
  num: number,
  arr: ReadonlyArray<[ElementType, Key]>
): string {
  return `${num}:${createArrayId(arr)}`;
}
