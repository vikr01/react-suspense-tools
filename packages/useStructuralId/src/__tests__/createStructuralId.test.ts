import * as React from 'react';
import {createArrayId, clear} from '../createStructuralId';

Object.defineProperty(global, '__DEV_STRUCTURAL_ID_DEBUG__', {
    value: false,
    writable: true,
    configurable: true,
});

describe('createStructuralId', ()=>{
    afterEach(()=>{
        clear();
    });
    
    function Component1() {return null};
    const Component2 = ()=>null;
    class Component3 extends React.Component {
        render() {
            return null;
        }
    }
    const Component4 = React.forwardRef(function Component4() {return null;});
    const Component5 = null;
    const Component6 = React.memo(()=>null);
    const Component7 = 'a' as const;
    const Component8 = null;

    it('genereates a string from elements with various element types', ()=>{
        const components = [
            Component1,
            Component2,
            Component3,
            Component4,
            Component5,
            Component6,
            Component7,
            Component8,
        ];

        const values = declareElementsArray(components.map((comp)=>[comp, 0]));

        const id = createArrayId(values);
        expect(id).toMatchInlineSnapshot(`"000000:n[0],000001:n[0],000002:n[0],000003:n[0],Unknown:n[0],000004:n[0],a:n[0],Unknown:n[0]"`);
    });

    it('properly reuses ids of component pointers', ()=>{
        const components = declareElementsArray([
            [Component2, 'foobar'],
            [Component3, 'bazfoo'],
            [Component4, 10],
            [Component7, 15],
        ]);

        const [component3uniqueId] = createArrayId(components).split(',')[1].split(':');

        expect(createArrayId([[Component3, 'xyz']])).toBe(`${component3uniqueId}:s[xyz]`);
    });

    it('makes readable ids when the flag is on', ()=>{
        Object.defineProperty(global, '__DEV_STRUCTURAL_ID_DEBUG__', {
            value: true,
            writable: true,
            configurable: true,
        });

        const components = declareElementsArray([
            [Component8, 503],
            [Component4, 'baz'],
            [Component6, 11555],
            [Component1, 'null'],
        ]);

        const id = createArrayId(components);
        
        expect(id).toMatchInlineSnapshot(`"Unknown:n[503],Fallback:s[baz],Fallback:n[11555],Component1:s[null]"`);
    });
});

function declareElementsArray<T extends Parameters<typeof createArrayId>[0]>(value: T): T {
    return value;
}