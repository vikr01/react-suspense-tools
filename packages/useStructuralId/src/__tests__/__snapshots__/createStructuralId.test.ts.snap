// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStructuralId genereates a string from elements with various element types 1`] = `"000000:n[0],000001:n[0],000002:n[0],000003:n[0],Unknown:n[0],000004:n[0],a:n[0],Unknown:n[0]"`;

exports[`createStructuralId makes readable ids when the flag is on 1`] = `"Unknown:n[503],Fallback:s[baz],Fallback:n[11555],Component0:s[null]"`;
